p2p/conn: FlushStop. Use in pex. Closes #2092

ebuchman opened the pull request:
In seed mode, we call StopPeer immediately after Send.
Since flushing msgs to the peer happens in the background,
the peer connection is often closed before the messages are
actually sent out. With FlushStop, we allow all msgs to first be written and flushed out
on the conn before it is closed.

This should be a proper solve for this issue, to replace https://github.com/tendermint/tendermint/pull/2797/files#diff-8c6b1a350a161ced01d68514ccda7484R224

Closes #2092 

<!-- Thanks for filing a PR! Before hitting the button, please check the following items.-->

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

melekes commented on p2p/pex/pex_reactor_test.go:
```diff
@@ -40,7 +40,7 @@ func TestPEXReactorBasic(t *testing.T) {
 }
 
-func TestPEXReactorAddRemovePeer(t *testing.T) {
+func TestPEXReactorAddRemoveneer(t *testing.T) {
```
typo?

```suggestion
TestPEXReactorAddRemovePeer
```

melekes commented on p2p/pex/pex_reactor.go:
```diff
@@ -221,6 +221,7 @@ func (r *PEXReactor) Receive(chID byte, src Peer, msgBytes []byte) {
 		// 2) limit the output size
 		if r.config.SeedMode {
 			r.SendAddrs(src, r.book.GetSelectionWithBias(biasToSelectNewPeers))
+			src.FlushStop()
```
so we're calling FlushStop() and after that Stop() in StopPeerGracefully()? Won't it make sense to have a Flush() func and leave stopping to OnStop?

pros:
- no code duplication
- separation of concerns

ebuchman replied:
The problem is, to ensure there's no races or new messages that get queued and not sent, we need the Flush func to stop the service so that all queued sends can be flushed and no new ones are concurrently created. I couldn't think of a better way to do it.

One alternative was to create a FlushSend method which waited until a send was actually flushed out on the conn, but that was more complicated than this solution.

melekes replied:
> there's no races or new messages that get queued and not sent, we need the Flush func to stop the service so that all queued sends can be flushed and no new ones are concurrently created.

Yeah, right. But why do you care about them. Here we only care about flushing one msg ^^^.

ebuchman replied:
Because we don't want to have a race condition in the code? We can't safely call `sendSomePacketMsgs` directly until the sendRoutine has exited otherwise we'll have a real race condition.

I agree with you in principle, but it would take some rearchitecting of the code to do what we want safely here

melekes commented on p2p/conn/connection.go:
```diff
@@ -200,15 +202,56 @@ func (c *MConnection) OnStart() error {
+func (c *MConnection) OnStop() {
+	select {
+	case <-c.quit:
+		// already quit via FlushStop
```
🤢 see my comment above

jaekwon commented on p2p/conn/connection.go:
```diff
@@ -200,15 +202,56 @@ func (c *MConnection) OnStart() error {
-// OnStop implements BaseService
-func (c *MConnection) OnStop() {
+// FlushStop replicates the logic of OnStop,
+// but flushes all pending msgs before closing the connection.
```
// This method ensures that all successful .Send() calls will get flushed before closing

jaekwon commented on p2p/conn/connection.go:
```diff
@@ -85,6 +85,7 @@ type MConnection struct {
 	config        MConnConfig
 
 	quit       chan struct{}
+	done       chan struct{}
```
I'd comment here that closing .quit "causes" .done to close.  Not obvious, but nice convention.

jaekwon commented on p2p/conn/connection.go:
```diff
@@ -200,15 +202,56 @@ func (c *MConnection) OnStart() error {
-func (c *MConnection) OnStop() {
+// FlushStop replicates the logic of OnStop,
+// but flushes all pending msgs before closing the connection.
+func (c *MConnection) FlushStop() {
```
Is this safe with concurrent/repeated FlushStop calls?  The Service Stop/OnStop mechanism ensures that OnStop only gets called once.  Not so here.

ebuchman commented on p2p/conn/connection.go:
```diff
@@ -200,15 +202,56 @@ func (c *MConnection) OnStart() error {
+// FlushStop replicates the logic of OnStop,
+// but flushes all pending msgs before closing the connection.
+func (c *MConnection) FlushStop() {
 	c.BaseService.OnStop()
```
No because we need to run some logic before the Conn is closed in OnStop

ebuchman commented:
@melekes @jaekwon addressed all feedback - PTAL.

I had to add some extra logic to ensure FlushStop is not called more than once.

Also realized the seed node was disconnecting from all peers as soon as it received a pex-request, rather than just inbound peers.
